---
title: "sc-testis-seurat_human-neonate"
author: "Brad Hansen"
date: "7/13/2022"
output: html_document
---
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load packages, include=FALSE}

library(Seurat)
library(dplyr)
library(patchwork)
library(metap)


if (!require("pacman")) install.packages("pacman")
pacman::p_load(mosaicData, tibble, tidyr, stringr, purrr)


```

## Load Seurat Objects

Each dataset will become a unique Seurat object. The data is loaded based on the: matrix, genes, barcodes files.

following steps based on [Seurat tutorial](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html)


```{r load data}

gen_direct = "/bigdata/faustmanlab/bch/sc-testis/human/gse124263"


h_pnd2.d <- Read10X(data.dir = paste0(gen_direct,"/pnd2/"))
h_pnd7.d <- Read10X(data.dir = paste0(gen_direct,"/pnd7/"))


```


```{r build object}

h_pnd2 <- CreateSeuratObject(counts = h_pnd2.d, project = "sc-testis-h", min.cells = 3, min.features = 200)
h_pnd7 <- CreateSeuratObject(counts = h_pnd7.d, project = "sc-testis-h", min.cells = 3, min.features = 200)



```

## Preprocess individual datasets

Now that the datasets are loaded as Seurat objects -- use the mitochondrial DNA and RNA counts metrics for initial QC.


THe first step is to identify the percentage of transcripts coming from mitochondria, this is indicative of cell death and may not be something we want to keep (ie remove dead cells)

```{r percent mitochondrial, echo=FALSE}

h_pnd2[["percent.mt"]] <- PercentageFeatureSet(h_pnd2, pattern = "^MT-")
h_pnd7[["percent.mt"]] <- PercentageFeatureSet(h_pnd7, pattern = "^MT-")



h_pnd2[["age_weeks"]] <- "pnd2"
h_pnd7[["age_weeks"]] <- "pnd7"

```

`Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

```{r filter, eval=FALSE, fig.show="hold", include=FALSE, out.width="50%"}}

vln1 <- VlnPlot(h_pnd2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
vln2 <- VlnPlot(h_pnd7, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)


```

```{R qc plots}

plot1 <- FeatureScatter(h_pnd2, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(h_pnd7, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

```
```{r subset objects}

h_pnd2 <- subset(h_pnd2, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
h_pnd7 <- subset(h_pnd7, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)


```


Now moving on to dataset integration based on [this workflow](https://satijalab.org/seurat/articles/integration_introduction.html)

```{r normalize}

sc_testis_h.neo.list <- c(h_pnd2, h_pnd7)


sc_testis_h.neo.list <- lapply(X = sc_testis_h.neo.list, FUN = SCTransform)



# select features that are repeatedly variable across datasets for integration
features_all <- SelectIntegrationFeatures(object.list = sc_testis_h.neo.list, nfeatures = 3000)
sc_testis_h.neo.list <- PrepSCTIntegration(object.list = sc_testis_h.neo.list, anchor.features = features_all)

```

```{r object integration}

testis.anchors_all <- FindIntegrationAnchors(object.list = sc_testis_h.neo.list, normalization.method = "SCT", anchor.features = features_all)
# this command creates an 'integrated' data assay
testis.combined_all <- IntegrateData(anchorset = testis.anchors_all, normalization.method = "SCT")

```

```{r analyze integrated data}

# original unmodified data still resides in the 'RNA' assay
DefaultAssay(testis.combined_all) <- "integrated"



# Run the standard workflow for visualization and clustering

testis.combined_all <- RunPCA(testis.combined_all, verbose = FALSE)
testis.combined_all <- RunUMAP(testis.combined_all, reduction = "pca", dims = 1:30)
testis.combined_all <- FindNeighbors(testis.combined_all, reduction = "pca", dims = 1:30)
testis.combined_all <- FindClusters(testis.combined_all, resolution = 0.5)

```


```{r plots}

p1 <- DimPlot(testis.combined_all, reduction = "umap", group.by = "age_weeks")
p2 <- DimPlot(testis.combined_all, reduction = "umap", label = TRUE, repel = TRUE)
p1 + p2

```



```{r saveloadRdatafile}

save(testis.combined_all, file = "/bigdata/faustmanlab/bch/sc-testis/human/gse124263/testis.pnd2pnd7_comb.RData")



#load(file = "/bigdata/faustmanlab/bch/sc-testis/human/gse124263/testis.pnd2pnd7_comb.RData")
```
## Next Use Seurat to identify conserved cell markers from the clusters across time

from tutorial "This function performs differential gene expression testing for each dataset/group and combines the p-values using meta-analysis methods from the MetaDE R package. For example, we can calculated the genes that are conserved markers irrespective of stimulation condition in cluster 6 (NK cells)."

this uses the original data, not the integrated data

```{r conserved markers}

DefaultAssay(testis.combined_all) <- "RNA"


clusts <- 0:11


conserved.markers_all <- lapply(X = clusts, FUN = function(x){
  
  x <- FindConservedMarkers(testis.combined_all, ident.1 = x, grouping.var = "age_weeks", verbose = TRUE)
  })


```

save as data files

```{r save files}

save(conserved.markers_all, file = "/bigdata/faustmanlab/bch/sc-testis/human/gse124263/conserved.markers_neonate_all.RData")




```


load data files

```{r load files}

#load(file = "/bigdata/faustmanlab/bch/sc-testis/human/gse124263/conserved.markers_neonate_all.RData")


```


Now we have Seurat objects for all data, embryonic only, and fetal only.

We also have the, "conserved.markers_all," 


```{r flatten list}

names(conserved.markers_all) <- clusts[0:11]


# Flatten list of dataframes


conserved.markers_allneonate_top10 <- map_df(conserved.markers_all, head, 10, .id = 'cluster')

write.csv(conserved.markers_allneonate_top10, file="/bigdata/faustmanlab/bch/sc-testis/human/gse124263/conserved.markers_allneonate_top10.csv")

```